backend/src/models/userModel.js

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-use-before-define */
+/* eslint-disable no-underscore-dangle */
+import mongoose from 'mongoose';
+import validator from 'validator';
+import bcrypt from 'bcrypt';
+import jwt from 'jsonwebtoken';
+
+const SALT_WORK_FACTOR = 10;
+
+// TODO: handle password, don't save token.
+const userSchema = new mongoose.Schema({
+  email: {
+    type: String,
+    unique: true,
+    required: true,
+    trim: true,
+    lowercase: true,
+    validate(value) {
+      if (!validator.isEmail(value)) {
+        throw new Error('Email is invalid');
+      }
+    },
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  tokens: [{
+    token: {
+      type: String,
+      required: true,
+    },
+  }],
+}, {
+  timestamps: true,
+});
+
+userSchema.methods.toJSON = function toJSON() {
+  const user = this;
+  const userObject = user.toObject();
+
+  delete userObject.password;
+  delete userObject.tokens;
+  delete userObject._id;
+  delete userObject.createdAt;
+  delete userObject.__v;
+
+  return userObject;
+};
+
+userSchema.statics.findByCredentials = async function findByCredentials(email, password) {
+  const user = await User.findOne({ email });
+
+  if (!user) {
+    throw new Error('User not found');
+  }
+
+  const isPasswordMatch = await User.comparePassword(password, user.password);
+
+  if (!isPasswordMatch) {
+    throw new Error('Credentials not right');
+  }
+  return user;
+};
+
+userSchema.pre('save', async function hashAndSavePassword(next) {
+  const user = this;
+  if (user.isModified('password')) {
+    user.password = await bcrypt.hash(user.password, SALT_WORK_FACTOR);
+  }
+  next();
+});
+
+userSchema.statics.comparePassword = async function comparePassword(candidatePassword, storedHash) {
+  const isMatch = await bcrypt.compare(candidatePassword, storedHash);
+  return isMatch;
+};
+
+userSchema.methods.generateAuthToken = async function generateAuthToken() {
+  const user = this;
+  // const timeInMilliseconds = Date.now();
+  const token = jwt
+    .sign({ _id: user._id.toString() }, process.env.JSON_WEB_SECRET_KEY);
```

Comment:
Import all the env variables from a constant/config file. Check this issue https://github.com/nodejs/node/issues/26213

Reply:
Done, now have a config file and environment variables are assigned once to const variables.